llvm/lib/CheerpUtils/PartialExecuter.cpp

```diff
@@ -0,0 +1,1421 @@
+				continue;
+
+			GenericValue SRC = getOperandValue(GV);
+			GenericValue *Ptr = (GenericValue*)GVTORP(SRC);
+			
+			validLoadIntervals.push_back({(long long)Ptr, ((long long)Ptr)+getDataLayout().getTypeAllocSize(GV->getInitializer()->getType())});
+		}
+
+		std::sort(validLoadIntervals.begin(), validLoadIntervals.end());
+	}
+	template <PartialInterpreter::VisitingPolicy policy>
+	void visitOuter(const llvm::Instruction& I) 
+	{
+		if (policy == REMOVE_VALUES)
+		{
+			removeFromMaps(&I);
+			return;
+		}
+		
+		if (const PHINode* phi = dyn_cast<PHINode>(&I))
+		{
+			if (const llvm::BasicBlock* from = getIncomingBB())
+			{
+				const llvm::Value* incomingVal = phi->getIncomingValueForBlock(from);
+				assert(incomingVal);
+				if (isValueComputed(incomingVal))
+				{
+					incomingsVal.push_back({phi, incomingVal});
+					return;
+				}
+			}
+			notComputedPhis.push_back(phi);
+			return;	
+		}
+		else
+		{
+			//While visiting the first NonPhi instruction we set the maps state correctly
+			for (auto& p : incomingsVal)
+				assignToMaps(p.first, p.second);
+			
+			for (auto& p : notComputedPhis)
+				   removeFromMaps(p);	
+			
+			//Then we clear the vectors since this has to be done only once
+			incomingsVal.clear();
+			notComputedPhis.clear();
+		}
+
+		//Here PHI have been properly processed
+
+		bool skip = hasToBeSkipped(I);
+
+		if (isBottomStack())
+		{
+			if (I.isTerminator())
+				skip = true;
+
+			//Skip Instructions we don't have enough information to execute
+			if (skip)
+				return;
+		}
+		else
+		{
+			//Terminators have special handling done here explicitly
+			if (I.isTerminator())
+			{
+				const BasicBlock* next = nullptr;
+
+				switch (I.getOpcode())
+				{
+					case Instruction::Br:
+					{
+						const BranchInst& BI = cast<BranchInst>(I);
+						
+						if (BI.isConditional())
+						{
+							if (isValueComputed(BI.getCondition()))
+							{
+								GenericValue V = getOperandValue(BI.getCondition());
+								next = BI.getSuccessor((V.IntVal == 0u) ? 1 : 0);
+							}
+						}
+						else
+							next = BI.getSuccessor(0);
+						break;
+					}
+					case Instruction::Switch:
+					{
+						const SwitchInst& SI = cast<SwitchInst>(I);
+				
+						if (isValueComputed(SI.getCondition()))
+						{
+							GenericValue V = getOperandValue(SI.getCondition());
+							ConstantInt* CI = (ConstantInt*)ConstantInt::get((IntegerType*)(SI.getCondition()->getType()), V.IntVal);
+							next = SI.findCaseValue(CI)->getCaseSuccessor();
+						}
+						break;
+					}
+					case Instruction::Ret:
+					{
+						const ReturnInst& RI = cast<ReturnInst>(I);
+						const Value* retVal = RI.getReturnValue();
+						const BitMask BITMASK = retVal ? getBitMask(RI.getReturnValue()) : NONE;
+						const Value* caller = getPenultimateStack().Caller;
+
+						Interpreter::visit(const_cast<Instruction&>(I));
+
+						stronglyKnownBits.pop_back();
+						if (retVal)
+							stronglyKnownBits.back()[caller] = BITMASK;
```

Comment:
This does not look correct. caller is the function, not the return values. Why are we setting the bitmask for the function itself?

Reply:
Or is caller the CallInst?

Reply:
Clarified